### PR TITLE
Practice in Tournaments! + Multi-entry bug fix

### LINKED ISSUE #369: Tournaments deduct more tickets.
This may be a logical bug.
When I quickly click on the button to join a tournament item,double the tickets deducted.

[bug trigger video](https://user-images.githubusercontent.com/65884968/158006594-75fd9e62-000c-4935-92ef-f43352ea63a5.mp4)

Platform: Windows 11
Bombsquad version: 1.6.10

Use Google translate, please forgive me if the grammar is wrong

### PR DESCRIPTION
**Pull request by 3alTemp**
## Description
As suggested in this [Discord post](https://discord.com/channels/1001896771347304639/1193869152495079511/1193869152495079511), this suggestion adds a practice button to tournaments that allow players to play tournaments without using tickets at the expense of not having their scores submitted.

Players are now able to normally run tournaments with ticket & medal awards, and practice tournaments while having their score skipped. When pressing the "restart" button mid-run, if on practice mode, the run will restart immediately, and if in a normal ticket run, they'll be prompted to buy another entry in order to start over, with no practice button available. If restarting from the score screen, they'll be prompted to choose between paid or practice. 

image

This idea tackles two very commonly reported issues:
* People wanting to play tournaments that have modified the game in minor ways or that have plugins, disabled or not, wasting their tickets because their client won't allow them to publish scores.
* People that are unable to play tournaments for any other reason (e.g., having an incompatible device, lack of tickets, etc.) having to use tickets to experience tournaments.

This works as a solid middle ground that allows a wide range of players to play tournaments without having to sacrifice tickets knowing their scores won't get submitted anyways, and for people that want to practice some runs before instead of burning tickets on each attempt or that don't have the modes available due to them being behind bspro or not having tickets, while letting competitive players participate as usual with no drawbacks.

The PR also makes some minor tweaks such as moving the close and "get tickets" buttons slightly closer to the window to prevent floating buttons and increasing the window initial transition-out delay to prevent buying multiple tournament entries.

## Type of Changes
|     | Type                   |
|-----|------------------------|
| ✓   | :bug: Bug fix          |
| ✓   | :sparkles: New feature |

## Related Issues
Closes #369

### CODE REVIEW

**Comment by 3alTemp** [comment]
Also half-joke half-serious, may I have a promo code that gives me like ~400 tickets? I wasted tons testing stuff trying to make everything work right x).

**Comment by CrossJoy** [comment]
what's the difference between tournament practice and practice (the bottom practice minigame list)?

**Comment by vishal332008** [comment]
> People wanting to play tournaments that have modified the game in minor ways or that have plugins, disabled or not, wasting their tickets because their client won't allow them to publish scores.

So you just have to make game check itself if it is modified or not before allowing the player to play and lose tickets.

> People that are unable to play tournaments for any other reason (e.g., having an incompatible device, lack of tickets, etc.) having to use tickets to experience tournaments.

I don't know about "incompatible device" (idk how bombsquad will even run in the first place). But lack of tickets, that is easy to get. Just join game everyday to get daily tickets. The minimum tickets to join a tournament is 5 or 10 (I didn't play tournaments in a long time. Sadly, i also got the modified problem). The game itself has practice rounds in different kinds of games with smart bots (annoying bots to be honest).

So I can't understand the exact reason why you are trying to implement the practice button. And restarting the game, it just makes the player to be careful of not making mistakes or wrong moves in the game. It was completely intentional by Eric.

**Comment by 3alTemp** [comment]
> what's the difference between tournament practice and practice (the bottom practice minigame list)?

1. The default practice modes are locked behind pro / tickets, which is odd by game design considering they're practice modes. I guess it's intentional to have some in-game monetization, so this could be used like a mode rotation. Efro mentioned coffees ago he wanted the monetization to be less aggressive and let players play more without the need of using tickets; one of the main reasons the pay-to-respawn system got removed, and one of the big reasons I'm suggesting this in the first place.
2. Race, Pro Race and it's sub-modes are actually unavailable at the practice tab in there while they're fully playable on tournaments, which might be an oversight to be honest x)
3. Tournaments have their own leaderboards compared to their menu practice counterparts, playing on their practice modes let's you rank up on performance along with publishing your score on it's leaderboards, while doing tournament practice completely discards your score as far as I know.

So players are still encouraged to get the normal practice modes so they can play them whenever they want to and to have their scores stored in the leaderboards, which would assist them on ranking up, which essentially means more tickets on log-in.

**Comment by 3alTemp** [comment]
> So you just have to make game check itself if it is modified or not before allowing the player to play and lose tickets.

That was the plan initially, but [era pointed out to me that Eric didn't want to put that particular check pre-game in case people wanted to pull out exploits, just to make them that more annoying to work with](https://discord.com/channels/1001896771347304639/1193869152495079511/1193897673489666208), so this sounded like the better next alternative.

> I don't know about "incompatible device".

People have been complaining for a while that their devices (phone, tv, etc.) have been detected as "rooted" despite them not having modified them in the first place at all. While I think efro doesn't have nearby plans to check that out, this at least would let those players not burn their tickets for tournaments over an apparent unfixable issue.

> Lack of tickets, that is easy to get. Just join game everyday to get daily tickets.

Let's not forget that players that are not on gold / diamond league get less tickets on launch, and that some of the entry prices range from 5 tickets to even 25, which is kind of an unbalance considering only the winner of the tournaments wins tickets back, and even that doesn't even guarantee a net win when the strat is wasting even more tickets for each attempt, plus let's not forget that non-paid players might also want to get a skin from the shop with the money they're getting, so at the end, for the average user, the tournament grind ends up being a living hell. Overall, the system has always been a little bit flawed, and *heavily* favors those who buy tickets.

> The game itself has practice rounds in different kinds of games.

Refer to the response above.

> I can't understand the exact reason why you are trying to implement the practice button. And taking same amount of tickets for restarting the game in a tournament, it just makes the player to be careful of not making mistakes or wrong moves in the game. It was completely intentional by Eric. For 1 game, this much tickets. If you want to play again, means you have to give tickets again. Basic game logic.

As I stated on the other response, I don't think efro wants to have this sort of systems anymore? Yeah, it's logical that each entry means you have to outperform the other players, and the practice button doesn't discredit that, you still have to do a ticket entry to submit a score and have to play good in order to win the tournament, the practice button's job is to give those other players who don't have 90k tickets to have the same strategic learning opportunity as the ones who do have 90k tickets, literally one of the main reasons why other games don't charge you to play in tournaments; imagine if Fortnite charged you like 200 vbucks EVERY TIME if you wanted to go competitive, that sounds ridiculous.

This addition would be a good step towards the right direction of not forcing players to burn tickets if they want to have a chance of playing the game, it's simple and it actually doesn't even change that much of how the game currently works.

**Comment by 3alTemp** [comment]
I have no idea what most of the pylint errors are talking about.

**Comment by efroemling** [comment]
So I'm a bit torn on this one. On one hand, for people taking the competition part of tournaments seriously, it is nice to be able to practice without having to look around for the same game in the practice row or mod the game or whatever. But on the other hand I feel that one mechanic of Tournaments (theoretically) is paying to play some 'special' variation of a game that you normally don't get to play. I know that's not really how things are right now since most all tournament games are available elsewhere, but I'd kinda prefer to keep that option open (gotta pay the bills and all that).

What if we did a compromise where practice mode is not available by default but can be enabled by manually setting a config value? Then it would keep casual players incentivized to pay for tournaments to 'try things out', but it would also allow competitive pro types to hone their skills and practice, and as a bonus would encourage people to learn how to hack around on the game a bit which I'd be happy about.

Thoughts?

**Comment by 3alTemp** [comment]
Well, one big reason that I should've probably mentioned at the root post on why I wanted to implement this is that, not only are people with no pro at a major disadvantage due to not being able to even use the "Challenges" variant (which is kind of crappy), but, as far as I'm aware, [some pro players who play tournaments still have tons of cheated tickets on their account from before v2](https://github.com/efroemling/ballistica/issues/300), completely jeopardizing the whole tickets and fairness system in a way for those who don't actively spend lots of money on tickets. Though, yeah, I guess it still brings a decent amount of income despite that issue, and considering how things currently are, I guess keeping it that way at least for the time being sounds fair.

But the compromise does sound like a good start! It does reward those people who constantly lurk and play around with code, letting them hop in practice runs without having to copy, try and delete plugins for each real runs they want to have, but it's still hidden enough for it to not become "mainstream".

**Comment by 3alTemp** [comment]
> Some checks were not successful

YES CI, I KNOW THE CHECK FAILED BECAUSE I MODIEFIED THE CHANGELOG ONLY. 😡

**Comment by efroemling** [comment]
Thanks!

**Comment by ZackerDC** [comment]
Hello, player tournaments is here 🐓, I would like to talk a little bit about this change.
First of all I want to mention that adding something like this change can make the "Practice" section at the end of all tournaments, which contains the levels you have paid tickets for, useless.
This feature should be available if you have unlocked the game mode or purchased it. Although I also suggest that some modes should appear in the original Practice section.
Race Big G Lake Frigit and Normal Football should be available (remember that if they are in the store it would mean removing them from the game lists, but an exception could be made).
Uber Soccer and Runaround Uber: You should practice them if you unlocked them in campaign mode (I also mention that they should appear in the original Practice section).
Infinite Onslaught - Runaround: Available to practice if you bought it BombSquad Pro
The last Stand: This mode is so competitive that I feel it should only be practiced in the original campaign and not from a tournament.

That would be somewhat counterproductive (if the modes of the tournaments as a whole are in the original practice) because you have 24/7 available the practice of all the tournaments in the practice section and this would remain as a shortcut button.

Also, if the above is not done, it makes it pointless to buy the modes for the practice section in the store since you can practice them for free when you enter a tournament.

Something to mention is that, if you have the power to practice a tournament infinitely, it can cause an imbalance in the tournaments, there are people with 100k tickets that they got using a trick of a modpack (AYMP) and you use those tickets to have practically infinite attempts which they use to practice the tournaments, in my opinion, this generation that continue to maintain that high number of tickets that makes them win tournaments in an easier way.

In conclusion, adding the practice button to the tournaments can be something necessary that implies adjusting several factors (adding missing modes and others) so that the practice section in which you buy levels does not become meaningless and it is no longer profitable to buy them.

One point I mentioned was that there were missing modes in the original Practice section, I hope they become available whether this is added or not (regardless if they are worth tickets of course).

Use translator, some things may not make sense, but I know what I mean (I think) 🦔.

**Comment by MatiasMGz** [comment]
Now that tournaments are free, i don't think practice would ever be needed.

**Review by Gamer1109j10 — APPROVED** [review]

**Review by Gamer1109j10 — APPROVED** [review]
